flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/logical/TemporalJoinRewriteWithUniqueKeyRule.scala

```diff
@@ -0,0 +1,139 @@
+  with CommonTemporalTableJoinRule {
+
+  override def matches(call: RelOptRuleCall): Boolean = {
+    val join = call.rel[FlinkLogicalJoin](0)
+    val snapshot = call.rel[FlinkLogicalSnapshot](2)
+    val snapshotInput = call.rel[FlinkLogicalRel](3)
+
+    val isTemporalJoin = matches(snapshot)
+    val canConvertToLookup = canConvertToLookupJoin(snapshot, snapshotInput)
+    val supportedJoinTypes = Seq(JoinRelType.INNER)
+
+    isTemporalJoin && !canConvertToLookup && supportedJoinTypes.contains(join.getJoinType)
+  }
+
+  override def onMatch(call: RelOptRuleCall): Unit = {
+    val join = call.rel[FlinkLogicalJoin](0)
+    val leftInput = call.rel[FlinkLogicalRel](1)
+    val snapshot = call.rel[FlinkLogicalSnapshot](2)
+
+    val joinCondition = join.getCondition
+
+    val newJoinCondition = joinCondition.accept(new RexShuttle {
+      override def visitCall(call: RexCall): RexNode = {
+        if (call.getOperator == TemporalJoinUtil.TEMPORAL_JOIN_CONDITION &&
+        isRowTimeTemporalTableJoin(snapshot)) {
+          val snapshotTimeInputRef = call.operands(0)
+          val rightTimeInputRef = call.operands(1)
+          val leftJoinKey = call.operands(2).asInstanceOf[RexCall].operands
+          val rightJoinKey = call.operands(3).asInstanceOf[RexCall].operands
+
+          val rexBuilder = join.getCluster.getRexBuilder
+          val primaryKeyInputRefs = extractPrimaryKeyInputRefs(leftInput, snapshot, rexBuilder)
+          if (primaryKeyInputRefs.isEmpty) {
+            throw new ValidationException("Event-Time Temporal Table Join requires both" +
+              s" primary key and row time attribute in versioned table," +
+              s" but no primary key found.")
+          }
+          TemporalJoinUtil.makeRowTimeTemporalJoinConditionCall(rexBuilder, snapshotTimeInputRef,
+            rightTimeInputRef, leftJoinKey, rightJoinKey, primaryKeyInputRefs.get)
+        }
+        else {
+          super.visitCall(call)
+        }
+      }
+    })
+    val rewriteJoin = FlinkLogicalJoin.create(
+      leftInput, snapshot, newJoinCondition, join.getJoinType)
+    call.transformTo(rewriteJoin)
+  }
+
+  private def extractPrimaryKeyInputRefs(
+      leftInput: RelNode,
+      snapshot: FlinkLogicalSnapshot,
+      rexBuilder: RexBuilder): Option[Seq[RexNode]] = {
+    val rightFields = snapshot.getRowType.getFieldList
+    val fmq = FlinkRelMetadataQuery.reuseOrCreate(snapshot.getCluster.getMetadataQuery)
+
+    val uniqueKeys = fmq.getUniqueKeys(snapshot.getInput())
+    val fields = snapshot.getRowType.getFieldList
+
+    if (uniqueKeys != null && uniqueKeys.size() > 0) {
+      val leftFieldCnt = leftInput.getRowType.getFieldCount
+      uniqueKeys
+        .filter(_.nonEmpty)
+        .map(_.toArray
+          .map(fields)
+          .map(f => rexBuilder.makeInputRef(
+            f.getType,
+            leftFieldCnt + rightFields.indexOf(f)))
+          .toSeq)
+        .toArray
+        .sortBy(_.length)
+        .headOption
```

Comment:
it's hard to read, it's better to add some comments and split them into a few steps